pages/matrix-files/[...id].js

```diff
@@ -0,0 +1,210 @@
+// node_modules
+import PropTypes from "prop-types";
+import Link from "next/link";
+// components
+import AlternateAccessions from "../../components/alternate-accessions";
+import Attribution from "../../components/attribution";
+import Breadcrumbs from "../../components/breadcrumbs";
+import { FileDataItems } from "../../components/common-data-items";
+import {
+  DataArea,
+  DataAreaTitle,
+  DataItemLabel,
+  DataItemValue,
+  DataPanel,
+} from "../../components/data-area";
+import DerivedFromTable from "../../components/derived-from-table";
+import DocumentTable from "../../components/document-table";
+import { EditableItem } from "../../components/edit";
+import { FileHeaderDownload } from "../../components/file-download";
+import JsonDisplay from "../../components/json-display";
+import ObjectPageHeader from "../../components/object-page-header";
+import PagePreamble from "../../components/page-preamble";
+import SeparatedList from "../../components/separated-list";
+// lib
+import buildAttribution from "../../lib/attribution";
+import buildBreadcrumbs from "../../lib/breadcrumbs";
+import {
+  requestDocuments,
+  requestFileSets,
+  requestFiles,
+} from "../../lib/common-requests";
+import errorObjectToProps from "../../lib/errors";
+import FetchRequest from "../../lib/fetch-request";
+import {
+  checkForFileDownloadPath,
+  convertFileDownloadPathToFilePagePath,
+} from "../../lib/files";
+import { isJsonFormat } from "../../lib/query-utils";
+
+export default function MatrixFile({
+  attribution,
+  matrixFile,
+  fileSet = null,
+  documents,
+  derivedFrom,
+  derivedFromFileSets,
+  fileFormatSpecifications,
+  referenceFiles,
+  isJson,
+}) {
+  return (
+    <>
+      <Breadcrumbs />
+      <EditableItem item={matrixFile}>
+        <PagePreamble>
+          <AlternateAccessions
+            alternateAccessions={matrixFile.alternate_accessions}
+          />
+        </PagePreamble>
+        <ObjectPageHeader item={matrixFile} isJsonFormat={isJson}>
+          <FileHeaderDownload file={matrixFile} />
+        </ObjectPageHeader>
+        <JsonDisplay item={matrixFile} isJsonFormat={isJson}>
+          <DataPanel>
+            <DataArea>
+              <FileDataItems
+                item={matrixFile}
+                fileSet={fileSet}
+              ></FileDataItems>
+            </DataArea>
+          </DataPanel>
+          <DataAreaTitle>Matrix Details</DataAreaTitle>
+          <DataPanel>
+            <DataArea>
+              <>
+                <DataItemLabel>First Dimension</DataItemLabel>
+                <DataItemValue>{matrixFile.dimension1}</DataItemValue>
+              </>
+              <>
+                <DataItemLabel>Second Dimension</DataItemLabel>
+                <DataItemValue>{matrixFile.dimension2}</DataItemValue>
+              </>
```

Comment:
You don’t need the `<>…</>` on these two label/value pairs. React is really confusing on this, but every JSX expression can return a maximum of one component. That one component can wrap many other components. So the return on line 51 needs the `<>…</>` because it otherwise would return both the `<Breadcrumbs>` _and_ `<EditableItem>` components — two is too many. So we wrap both of those in one `<>…</>`. The `<DataArea>` above here on line 74 can be one of many because it’s wrapped by one `<DataPanel>` which itself can be many wrapped by `<JsonDisplay>`, which itself can be one of many wrapped by `<EditableItem>`, which _now_ can be one of many because it’s wrapped by the single `<>…</>`.

The `referenceFiles` expression starts a new JSX expression. You can’t have both `<DataItemLabel>` _and_ `<DataItemValue>` because that’s two. So they get wrapped by one `<>…</>`.

The `<>…</>` you use above don’t help (neither do they hurt, but it’s not standard JSX) because they already have the `<DataArea>` wrapper which can contain as many components as you like.

It’s weird, and the early days of React saw lots of complaints about this, especially since `<>…</>` didn’t exist, so you had to use HTML-generating `<div>` or `<span>` to wrap. But now people are used to it, and `<>…</>` generates no HTML — it just satisfies the syntax. Written Korean has a rule that’s kind of similar, which I found fascinating.